_WD_FrameList

mlipok opened the pull request:
## Pull request

### Proposed changes

was discussed here:
https://github.com/Danp2/au3WebDriver/issues/354

### Checklist

- [x] I have read and noticed the [CODE OF CONDUCT](https://github.com/Danp2/au3WebDriver/blob/master/docs/CODE_OF_CONDUCT.md) document
- [x] I have read and noticed the [CONTRIBUTING](https://github.com/Danp2/au3WebDriver/blob/master/docs/CONTRIBUTING.md) document
- [ ] I have added necessary documentation or screenshots (if appropriate)

### Types of changes

- [ ] Bugfix (change which fixes an issue)
- [x] Feature (change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (functional, structural)
- [ ] Documentation content changes
- [ ] Other (please describe)

### What is the current behavior?

no way to list frames list

### What is the new behavior?

it is easy with `_WD_FrameList`

### Additional context
https://github.com/Danp2/au3WebDriver/issues/354#issuecomment-1203042673

can be tested with:
```autoit
Func UserTesting() ; here you can replace the code to test your stuff before you ask on the forum

	_WD_Navigate($sSession, 'https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe')
	_WD_LoadWait($sSession)

	; Example 1 - from 'https://www.w3schools.com' get frame list as string
	ConsoleWrite(_WD_FrameList($sSession,  False) & @CRLF)

	; Example 2 - from 'https://www.w3schools.com' get frame list as array, with HTML content of each document
	Local $aFrameList = _WD_FrameList($sSession,  True, '', True)
	_ArrayDisplay($aFrameList, 'Example 2', 0, 0, Default, 'Absolute Identifiers > _WD_FrameEnter|Relative Identifiers > _WD_FrameEnter|IFRAME attributes|URL|Body ElementID|HTML')


	_WD_FrameEnter($sSession, 2)
	_WD_FrameEnter($sSession, 0)
	; Example 3 - from 'https://www.w3schools.com' get frame list as array, with relative location
	$aFrameList = _WD_FrameList($sSession, True, '', False)
	_ArrayDisplay($aFrameList, 'Example 3', 0, 0, Default, 'Absolute Identifiers > _WD_FrameEnter|Relative Identifiers > _WD_FrameEnter|IFRAME attributes|URL|Body ElementID|HTML')

	_WD_Navigate($sSession, 'https://stackoverflow.com/questions/19669786/check-if-element-is-visible-in-dom')
	_WD_LoadWait($sSession)

	; Example 4 - from 'https://stackoverflow.com' get frame list as string
	ConsoleWrite(_WD_FrameList($sSession,  False) & @CRLF)

	; Example 5 - from 'https://stackoverflow.com' get frame list as array
	#TODO this following line - NOT WORKS - should be checked
;~ 	$aFrameList = _WD_FrameList($sSession, True)
	$aFrameList = _WD_FrameList($sSession,  True, '', False)
	_ArrayDisplay($aFrameList, 'Example 5', 0, 0, Default, 'Absolute Identifiers > _WD_FrameEnter|Relative Identifiers > _WD_FrameEnter|IFRAME attributes|URL|Body ElementID|HTML')

	; Example 6 - from 'https://stackoverflow.com' get frame list as array - but only this one which contain id="question-header"
	$aFrameList = _WD_FrameList($sSession,  True, '(?i)id=.question-header.')
	_ArrayDisplay($aFrameList, 'Example 6', 0, 0, Default, 'Absolute Identifiers > _WD_FrameEnter|Relative Identifiers > _WD_FrameEnter|IFRAME attributes|URL|Body ElementID|HTML')

EndFunc   ;==>UserTesting
```

![image](https://user-images.githubusercontent.com/11089482/182446618-ebd5140e-4578-4920-90a6-888c3381e355.png)

### System under test
not related

mlipok commented:
big thanks to @TheDcoder  for his answer here:
https://www.autoitscript.com/forum/topic/205553-webdriver-udf-help-support-iii/?do=findComment&comment=1505451

It helps me to solve the problem with checking current frame level.

mlipok commented:
@Danp2  as you can see finally I decide to return as result both absolute and relative paths, so that I met mine and yours requirements without having to use a switch as a parameter.

mlipok commented on wd_helper.au3:
```diff
@@ -636,6 +636,156 @@ Func _WD_FrameLeave($sSession)
+
+	; Back to "calling frame"
+	_WD_FrameEnter($sSession, $sStartBody_ElementID)
+	#TODO decide if Error handling here is required or not ?
```
@Danp2  what you think abou Error handling here ?

mlipok replied:
Please take a look on: [error handling + headers](https://github.com/Danp2/au3WebDriver/pull/362/commits/74f6fb0c88c2cfd426b9e10b6604731e77b77c03)

> I would think that you need some form of error checking here because _WD_FrameEnter can generate an error.

suplemented

> I'm confused by this implementation. Why would you be calling _WD_FrameEnter with the body element as a target?

fixed

Danp2 commented:
I have not analyzed or run this code, but a major problem I can foresee is that of maintaining the current browsing context. If I am currently located two levels deep inside frames when calling a function, then I would expect to still be in that same "location" upon the function's return. 

Have you given this any consideration? If yes, then how do you plan to address it?

mlipok commented:
> I can foresee is that of maintaining the current browsing context. If I am currently located two levels deep inside frames when calling a function, then I would expect to still be in that same "location" upon the function's return.
> 
> Have you given this any consideration? If yes, then how do you plan to address it?

Yes of course, I just fixed this part, as you pointed me:
> I'm confused by this implementation. Why would you be calling _WD_FrameEnter with the body element as a target?

Take a look here:
https://github.com/Danp2/au3WebDriver/blob/74f6fb0c88c2cfd426b9e10b6604731e77b77c03/wd_helper.au3#L697-L700

mlipok commented:
here are my examples:

mlipok commented on wd_helper.au3:
```diff
@@ -636,6 +636,173 @@ Func _WD_FrameLeave($sSession)
+	If @error Then
+		$iErr = $_WD_ERROR_GeneralError
+	Else
+		$vResult = __WD_FrameList_Internal($sSession, 'null', '', $sFilter, $bReturnHTML)
```
`$sFilter` should not be maintained internally because then we loose `$sStartLocation` setting feature and for this reason

https://github.com/Danp2/au3WebDriver/blob/74f6fb0c88c2cfd426b9e10b6604731e77b77c03/wd_helper.au3#L697-L700

will be omnited

Danp2 replied:
Are you saying that you no longer intend to restore the browsing context before exiting?

mlipok replied:
No. I say that even if `$sFilter` is used it should back and for this reason I need to refactor this by move `$sFilter` functionality out of `__WD_FrameList_Internal` to `_WD_FrameList`.

btw.
WIP

mlipok replied:
fixed.
Check current version

Danp2 commented:
Here's the code for a very simple test that I ran --

```autoit
	_WD_Navigate($sSession, 'https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe')
	_WD_LoadWait($sSession)

	_WD_FrameEnter($sSession, 0)
	ConsoleWrite("IsWindowTop = " & _WD_IsWindowTop($sSession) & @CRLF)

	; Example 1 - from 'https://www.w3schools.com' get frame list as string
	ConsoleWrite(_WD_FrameList($sSession, False) & @CRLF)

	ConsoleWrite("IsWindowTop = " & _WD_IsWindowTop($sSession) & @CRLF)
	return
```

<details><summary>Results</summary>
<p>

```
+ wd_demo.au3: Running: UserTesting
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Navigate ==> Success [0] : Parameters:   URL=https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe
_WD_LoadWait ==> Success [0] : Parameters:    Delay=Default    Timeout=Default    Element=Default
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_IsWindowTop ==> Success [0]
IsWindowTop = False
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":null}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_GetFrameCount ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":null}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_GetFrameCount ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":null}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_GetFrameCount ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":null}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":0}
_WD_FrameEnter ==> Success [0] : Parameters:    Identifier=0
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_GetFrameCount ==> Success [0]
__WD_FrameList_Internal ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=parent   Option={}
_WD_FrameLeave ==> Success [0]
__WD_FrameList_Internal ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=parent   Option={}
_WD_FrameLeave ==> Success [0]
__WD_FrameList_Internal ==> Success [0]
__WD_Post ==> Success [0] : HTTP status = 200
_WD_Window ==> Success [0] : Parameters:   Command=parent   Option={}
_WD_FrameLeave ==> Success [0]
__WD_FrameList_Internal ==> Success [0]
__WD_Post ==> No match [8] : HTTP status = 404
_WD_Window ==> No match [8] : Parameters:   Command=frame   Option={"id":{"element-6066-11e4-a52e-4f735466cecf":"null/0"}}
_WD_FrameEnter ==> Webdriver Exception [10] : Parameters:    Identifier=null/0
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=False   Filter=   ReturnHTML=False Was not able to check / back to "calling frame"
null|null||https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe|9dc14e90-b58b-40ab-be92-ecd1dd65a3b3|
null/0||<iframe id="iframeResult" name="iframeResult" allowfullscreen="true" frameborder="0"></iframe>|https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe|15069330-e0f9-4795-a5de-df31054a7d77|
null/0/0|0|<iframe src="https://www.w3schools.com" title="W3Schools Free Online Web Tutorials">
</iframe>|https://www.w3schools.com/|4bd19bb2-7460-46f8-a256-96ebe875303e|
null/0/0/0|0/0|<iframe src="/howto/tryhow_js_slideshow_ifr.htm" id="howto_iframe"></iframe>|https://www.w3schools.com/howto/tryhow_js_slideshow_ifr.htm|8b09ff58-ba52-4b85-944f-92a3504ec44b|
__WD_Post ==> Success [0] : HTTP status = 200
_WD_ExecuteScript ==> Success [0]
_WD_IsWindowTop ==> Success [0]
IsWindowTop = True
+ wd_demo.au3: Finished: UserTesting
__WD_Delete ==> Success [0] : HTTP status = 200
_WD_DeleteSession ==> Success [0] : WebDriver session deleted
```

</p>
</details>

#### Comments

- Logs should be less verbose when in "Info" mode
- Some errors are occurring with _WD_FrameEnter due to `Identifier=null/0`. I'm guessing this is due to pending changes to _WD_FrameEnter that aren't part of this PR
- `IsWindowTop = False` before calling your function but True afterwards

mlipok commented:
> #### Comments
> * Logs should be less verbose when in "Info" mode

fixed in my WIP version

> * Some errors are occurring with _WD_FrameEnter due to `Identifier=null/0`. I'm guessing this is due to pending changes to _WD_FrameEnter that aren't part of this PR

exaclty

> * `IsWindowTop = False` before calling your function but True afterwards

WIP

mlipok commented:
> > #### Comments
> > 
> > * Logs should be less verbose when in "Info" mode
> 
> fixed in my WIP version

fixed on [further changes](https://github.com/Danp2/au3WebDriver/pull/362/commits/9f0418c87778c7bef6d23796ac1ef217be731fd1)

> > * `IsWindowTop = False` before calling your function but True afterwards
> 
> will check when I end refactoring `$sFilter`

uisng this commit: [further changes](https://github.com/Danp2/au3WebDriver/pull/362/commits/9f0418c87778c7bef6d23796ac1ef217be731fd1)

and 
```autoit
	$_WD_DEBUG = $_WD_DEBUG_Error
	ConsoleWrite("> " & @ScriptLineNumber & " IsWindowTop = " & _WD_IsWindowTop($sSession) & @CRLF)
	_WD_FrameEnter($sSession, 'null/2')
	If @error Then
		If @error Then MsgBox($MB_OK + $MB_TOPMOST + $MB_ICONERROR, "'null/2/0'", "@error = " & @error & @CRLF & "@extended = " & @extended)
		_WD_FrameEnter($sSession, Null)
		_WD_FrameEnter($sSession, 2)
	EndIf
	ConsoleWrite("> " & @ScriptLineNumber & " IsWindowTop = " & _WD_IsWindowTop($sSession) & @CRLF)

	; Example 7 - from 'https://stackoverflow.com' get frame list as array, with relative location
	$aFrameList = _WD_FrameList($sSession, True, '', False)
	_ArrayDisplay($aFrameList, 'Example 7v1 - relative to "null/2"', 0, 0, Default, 'Absolute Identifiers > _WD_FrameEnter|Relative Identifiers > _WD_FrameEnter|IFRAME attributes|URL|Body ElementID|HTML')
	ConsoleWrite("> " & @ScriptLineNumber & " IsWindowTop = " & _WD_IsWindowTop($sSession) & @CRLF)

	$aFrameList = _WD_FrameList($sSession, True, '', False)
	_ArrayDisplay($aFrameList, 'Example 7v2 - should still be relative to "null/2"', 0, 0, Default, 'Absolute Identifiers > _WD_FrameEnter|Relative Identifiers > _WD_FrameEnter|IFRAME attributes|URL|Body ElementID|HTML')
	ConsoleWrite("> " & @ScriptLineNumber & " IsWindowTop = " & _WD_IsWindowTop($sSession) & @CRLF)
```

I get this results:
```
> 970 IsWindowTop = True
> 977 IsWindowTop = False
> 982 IsWindowTop = False
> 986 IsWindowTop = False

```

mlipok commented on wd_helper.au3:
```diff
@@ -636,6 +636,199 @@ Func _WD_FrameLeave($sSession)
 EndFunc   ;==>_WD_FrameLeave
 
+; #FUNCTION# ====================================================================================================================
+; Name ..........: _WD_FrameList
```
@Danp2 please reaview `header` and in code `comments` for `_WD_FrameList`

Danp2 replied:
This will get done in due time

mlipok commented on wd_helper.au3:
```diff
@@ -636,6 +636,199 @@ Func _WD_FrameLeave($sSession)
+EndFunc   ;==>_WD_FrameList
+
+; #INTERNAL_USE_ONLY# ===========================================================================================================
+; Name ..........: __WD_FrameList_Internal
```
@Danp2 please reaview `header` and in code `comments` for `__WD_FrameList_Internal`

mlipok commented on wd_helper.au3:
```diff
@@ -636,6 +636,199 @@ Func _WD_FrameLeave($sSession)
+	Local $iErr = $_WD_ERROR_Success
+	Local $vResult = '', $s_URL = '', $s_HTML = '', $sMessage = ''
+
+	Local $a_Level = StringSplit($sLevel, '/')
```
as one of further step this part can be replaced to `_WD_FrameEnter($sSession, $sLevel)`

but after rethinking at first place this 
https://github.com/Danp2/au3WebDriver/pull/363
and 
https://github.com/Danp2/au3WebDriver/pull/364

(in the given order)

PR should be reviewed and merged

mlipok replied:
suplemented in:
[further changes](https://github.com/Danp2/au3WebDriver/pull/362/commits/3bdf5437e1f1792e68f9d69804f593ce4cd6ac52)

Danp2 commented on wd_helper.au3:
```diff
@@ -636,6 +636,199 @@ Func _WD_FrameLeave($sSession)
 
+; #FUNCTION# ====================================================================================================================
+; Name ..........: _WD_FrameList
+; Description ...: get list of documents (window.top and all frames) from current webdriver context/tab
```
A few comments about this --

- If it returns a list of documents, then I think the function should be renamed to something that better describes the action being performed
- It isn't clear to me what you mean by "from current webdriver context/tab". Does this function return a listing of all frames for the current tab or only those from the current browsing context?

mlipok replied:
> * If it returns a list of documents, then I think the function should be renamed to something that better describes the action being performed

_WD_GetFrames ?

Danp2 replied:
No. For me, the issue is that it seems like you are using "documents" in the description as if it were interchangeable with "frames", when they are two different things.

mlipok replied:
so maybe like this:
`; Description ...: get list of frames (window.top and all frames) from current webdriver context`

mlipok replied:
I know that `window.top` is not frame but WebDriver treat it like frames when it comes to Frame Entering as using NULL you are leaving all frames and you are on top but you must use "FRAME" function to do this.

mlipok replied:
better description proposal:
`; Description ...: list detailed information about frames possible to use (window.top and all frames)`

mlipok replied:
suplemented in:
[further changes](https://github.com/Danp2/au3WebDriver/pull/362/commits/3bdf5437e1f1792e68f9d69804f593ce4cd6ac52)

Danp2 replied:
>  ; Description ...: list detailed information about frames possible to use ("window.top" and all frames) 

Some people may not be familiar with "window.top". Here's my take on how this should read --

` ; Description ...: Retrieves a detailed list of the main document and all associated frames`

mlipok replied:
Please take a look here: [const parameters + header](https://github.com/Danp2/au3WebDriver/pull/362/commits/6276ff5011d2a1cc604dbffceb19079f7100ac47)

Danp2 commented on wd_helper.au3:
```diff
@@ -636,6 +636,199 @@ Func _WD_FrameLeave($sSession)
+; Author ........: mLipok
+; Modified ......:
+; Remarks .......: using $sFilter will set $bReturnHTML
+; Related .......: __WD_FrameList_Internal, _WD_FrameEnter, _WD_GetFrameCount
```
Related function list shouldn't include internal functions.

mlipok replied:
fixed in:
[further changes](https://github.com/Danp2/au3WebDriver/pull/362/commits/3bdf5437e1f1792e68f9d69804f593ce4cd6ac52)

Danp2 commented on wd_helper.au3:
```diff
@@ -636,6 +636,199 @@ Func _WD_FrameLeave($sSession)
+	Local $vResult = '', $iErr = $_WD_ERROR_Success, $iExt = 0
+	Local $sStartBody_ElementID = _WD_ExecuteScript($sSession, "return window.document.body;", Default, Default, $_WD_JSON_Element)
+
+	If $sFilter <> '' Then $bReturnHTML = True
+	If @error Then
+		$iErr = $_WD_ERROR_GeneralError
```
What if the user explicitly supplies a value of `False` for $bReturnHTML? It seems wrong to me to ignore their request. Also, I would generally expect the `@error` line to immediately follow the command that generated the error.

mlipok replied:
I used `$bReturnHTML_Internally`

Please take a look:
[further changes](https://github.com/Danp2/au3WebDriver/pull/362/commits/3bdf5437e1f1792e68f9d69804f593ce4cd6ac52)

Danp2 commented on wd_helper.au3:
```diff
@@ -636,6 +636,199 @@ Func _WD_FrameLeave($sSession)
+		; create array of frames from string returned from __WD_FrameList_Internal
+		_ArrayAdd($a_Result, $vResult)
+
+		; find "calling frame" - set $sStartLocation
+		For $i = 0 To UBound($a_Result) - 1
+			If $a_Result[$i][4] = $sStartBody_ElementID Then $sStartLocation = $a_Result[$i][0]
+		Next
+
+		; recalculate absolute path COL0 to relative path COL1
+		For $i = 0 To UBound($a_Result) - 1
+			$a_Result[$i][1] = StringRegExpReplace($a_Result[$i][0], '\A' & $sStartLocation & '\/?', '')
+		Next
```
This could be combined into a single loop, right?

mlipok replied:
Thanks.
After looking at this point I made further optimisation:

`Find "calling frame"` + `recalculate locations` + `apply filter` will be performed together within single loop.

Danp2 commented on wd_helper.au3:
```diff
@@ -636,6 +636,199 @@ Func _WD_FrameLeave($sSession)
+		_ArrayAdd($a_Result, $vResult)
+
+		; find "calling frame" - set $sStartLocation
+		For $i = 0 To UBound($a_Result) - 1
+			If $a_Result[$i][4] = $sStartBody_ElementID Then $sStartLocation = $a_Result[$i][0]
+		Next
+
```
You will need to convert "null" to Null (the autoit keyword) here. Otherwise, the _WD_FrameEnter in the `Back to "calling frame"` will fail if you were at the top level when calling this function.

mlipok replied:
I manage this other way.
Please take a look to **_WD_FrameEnter($sSession, 'null/2/0') - path support** here ['null' support + comments](https://github.com/Danp2/au3WebDriver/pull/364/commits/e6458b9acb7a02e6128865e6422ad97977638328)

Danp2 commented on wd_helper.au3:
```diff
@@ -638,40 +638,41 @@ EndFunc   ;==>_WD_FrameLeave
 ; Example .......: No
 ; ===============================================================================================================================
-Func _WD_FrameList($sSession, $bReturnAsArray = True, $sFilter = '', $bReturnHTML = False)
+Func _WD_FrameList(Const $sSession, Const $bReturnAsArray = True, Const $sFilter = '', Const $bReturnHTML = False)
```
Using `Const` on a function's parameters isn't commonly done, even when you don't intend to modify the parameter's value. So why do it here?

mlipok replied:
I tried whether it would work, and I wanted to know your opinion on this.

Ok.
Soon I revert the `const` change from `_WD_FrameLinst` parameters.

mlipok replied:
Please take a look here: [const parameters + header](https://github.com/Danp2/au3WebDriver/pull/362/commits/6276ff5011d2a1cc604dbffceb19079f7100ac47)

mlipok commented:
Now I have a question.
Would you like to see here an option to filter with `$_WD_LOCATOR_ByXPath` and `$_WD_LOCATOR_ByCSSSelector` and not only with  `RegExp pattern` ?

mlipok commented:
> Now I have a question. Would you like to see here an option to filter with `$_WD_LOCATOR_ByXPath` and `$_WD_LOCATOR_ByCSSSelector` and not only with `RegExp pattern` ?

I hope to allow the user to check on which FRAME the given element exists.

Danp2 commented:
> Would you like to see here an option to filter with $_WD_LOCATOR_ByXPath and $_WD_LOCATOR_ByCSSSelector and not only with RegExp pattern ?

I would say no. 

> I hope to allow the user to check on which FRAME the given element exists.

Which element?  And how would you intend to do that?  Are we still talking about _WD_FrameList? 😕

mlipok commented:
Never mind (about filtering with $_WD_LOCATOR_By....).

After rethinking I think that currently `_WD_FrameList` is even more complex then it should be with `$sFilter` and `$bReturnHTML` feature.

We can even remove those 2 parameters.
I think I can provide another separate `_WD_FrameListEx` to do this filtering stuff, and submit them on the forum here: https://www.autoitscript.com/forum/topic/201537-webdriver-example-scripts-collection/

mlipok commented:
As to [$sFrameAttributes new line removal](https://github.com/Danp2/au3WebDriver/pull/362/commits/959ee88ef87c8f315959870b7fb7b2aa9f1669b9)

Before:
![image](https://user-images.githubusercontent.com/11089482/183380829-7e37f125-360a-4a3f-8283-7b202af2a272.png)

after:
![image](https://user-images.githubusercontent.com/11089482/183380876-2b5b3b2c-d448-4198-8451-769beba5e814.png)

mlipok commented on wd_helper.au3:
```diff
@@ -636,6 +644,202 @@ Func _WD_FrameLeave($sSession)
+						$s_HTML = ''
+						$sMessage = 'Error occured on "' & $sLevel & '" level when getting HTML Source'
+					Else
+						$s_HTML = StringToBinary($s_HTML, $SB_UTF8)
```
Should we use `StringToBinary($s_HTML, $SB_UTF8)` or other for example: `$SB_ANSI` or even `__WD_Base64Encode` and `__WD_Base64Decode`

?

Danp2 replied:
Yes, something should be done here because it doesn't make sense to return the HTML as a binary string. Did you mean BinaryToString?

mlipok replied:
Please check: 
[BinaryToString usage and HTML CleanUp](https://github.com/Danp2/au3WebDriver/pull/362/commits/a65e8c960b1ff8bf5c4a47d5094da76d8ca29dae)

mlipok replied:
> Should we use `StringToBinary($s_HTML, $SB_UTF8)` or other for example: `$SB_ANSI`

I still wonder how character encoding on web pages follows and whether we can encode source HTML to BINARY with UTF8 or is it better to use ANSI ?

Danp2 replied:
It is a complex issue that could be avoided if you followed through on your recent statement 

> After rethinking I think that currently _WD_FrameList is even more complex then it should be with $sFilter and $bReturnHTML feature.

Danp2 replied:
I'm not likely to use this function, so I may not be the best judge of its functionality. I'll leave it to you to decide on the future of those features. If you decide to leave them in, then you should look at using `$_WD_BFORMAT` to control the binary formatting.

mlipok replied:
Please take a look on: [Removal of $sFilter and $bReturnHTML](https://github.com/Danp2/au3WebDriver/pull/362/commits/83d6f4c5f7a04a505ff14aa98521233cb711820c)

mlipok replied:
you can check it with modified wd_demo.au3: [DemoFrames _WD_FrameList and _WD_FrameEnter+path](https://github.com/Danp2/au3WebDriver/pull/362/commits/4073ade1b5581167ccd37de2f92d0585ca89b22b)

mlipok commented on wd_helper.au3:
```diff
@@ -672,6 +679,158 @@ Func _WD_FrameLeave($sSession)
+	EndIf
+	If $iErr Then $iErr = $_WD_ERROR_Exception
+
+	$sMessage = ($sMessage And $_WD_DEBUG = $_WD_DEBUG_Full) ? (' Information: ' & $sMessage) : ("")
```
I still propose here:
$sMessage = ($sMessage And $_WD_DEBUG > $_WD_DEBUG_Error) ? (' Information: ' & $sMessage) : ("")

Danp2 replied:
That should be ok.

Danp2 replied:
You could finish this up at the same time. 😉

mlipok replied:
Take a look on: [$sMessage on > $_WD_DEBUG_Error](https://github.com/Danp2/au3WebDriver/pull/362/commits/acb79955b41f168029ed70b5065a8acbb759ee6a)

Danp2 commented on wd_helper.au3:
```diff
@@ -683,96 +682,58 @@ EndFunc   ;==>_WD_FrameLeave
-;                  $sFilter             - [optional] RegExp pattern to filter frames with their HTML content. Default is ''.
-;                  $bReturnHTML         - [optional] Return frame document's HTML source? Default is False.
-; Return values .: Success - 2D array (with 6 cols) or string ( delimited with | and @CRLF )
+; Return values .: Success - 2D array (with 5 cols) or string ( delimeted with | and @CRLF )
```
"delimited" is misspelled

mlipok replied:
Thanks.
FIxed: [typo delimited](https://github.com/Danp2/au3WebDriver/pull/362/commits/66b70efcad58f9bd8847a9d925760a330ebed3ff)

Danp2 commented on wd_helper.au3:
```diff
@@ -683,96 +682,58 @@ EndFunc   ;==>_WD_FrameLeave
-				$vResult = _ArrayToString($a_Result)
-				If @error Then $vResult = ''
-			EndIf
+			$vResult = _ArrayToString($a_Result)
+			If @error Then $vResult = ''
```
Shouldn't $iErr and $sMessage get updated here since this affects the data returned if _ArrayToString fails?

mlipok replied:
Please take a look:
[added $_WD_ERROR_RetValue](https://github.com/Danp2/au3WebDriver/pull/362/commits/00a0e77562061316eec4a5670595ad40bcaec3d1)
and
[header suplemented](https://github.com/Danp2/au3WebDriver/pull/362/commits/a95a827e941555d30e6e24885155e11e41932271)

mlipok commented:
Here we should consider:
`Is there excessive logging that can be reduced by internally adjusting and resetting the value of $_WD_DEBUG within the function?`

mlipok commented:
> Here we should consider: `Is there excessive logging that can be reduced by internally adjusting and resetting the value of $_WD_DEBUG within the function?`

What is your opinion about this ?

Danp2 commented:
It should be fine if you perform the same revision here.

mlipok commented:
> It should be fine if you perform the same revision here.

Please take a look on: [__WD_FrameList_Internal - prevent multiple logging](https://github.com/Danp2/au3WebDriver/pull/362/commits/77cddd2c12999802f79dbe7ea1609f24b9a59041)

mlipok commented:
I think that we can strip the to:
```
__WD_FrameList_Internal ==> Success [0] : Parameters:    Level=null
_WD_FrameList ==> Success [0] : Parameters:    ReturnAsArray=False
```

btw.
I also have one other idea to check.

Danp2 commented:
> Please take a look on: [__WD_FrameList_Internal - prevent multiple logging](https://github.com/Danp2/au3WebDriver/pull/362/commits/77cddd2c12999802f79dbe7ea1609f24b9a59041)

I would have expected this change to be performed on the main _WD_Frame function instead.

mlipok commented on wd_helper.au3:
```diff
@@ -672,6 +679,177 @@ Func _WD_FrameLeave($sSession)
+	Local $vResult = '', $s_URL = '', $sMessage = ''
+
+	_WD_FrameEnter($sSession, $sLevel)
+	If $sLevel = 'null/3' Then SetError($_WD_ERROR_Exception)
```
this is only for testing should be removed before merging

mlipok replied:
Take a look on: [intentional testing error removed](https://github.com/Danp2/au3WebDriver/pull/362/commits/14ff948ec2734e80027648b2136c6e53fa54b25e)

Treat this part of conversationa as resolved.

mlipok commented:
> > Please take a look on: [__WD_FrameList_Internal - prevent multiple logging](https://github.com/Danp2/au3WebDriver/pull/362/commits/77cddd2c12999802f79dbe7ea1609f24b9a59041)
> 
> I would have expected this change to be performed on the main _WD_Frame function instead.

take a look on: [$_WD_DEBUG_Saved refactoring](https://github.com/Danp2/au3WebDriver/pull/362/commits/86a0a2225511268623b1b8b2aac76d0be36c1a8c)

I get such results:
```
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null/3 Information: Error occured on "null/3" level when trying to entering frame
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=False Information:  Was not able to check / back to "calling frame"
```

I can do it this way:
```
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null/3 Information: Error occured on "null/3" level when trying to entering frame
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=False Information:  Was not able to check / back to "calling frame"
```

mlipok commented:
here:

> btw.
> I also have one other idea to check.


and here:

> I can do it this way:
> 
> ```
> __WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null/3 Information: Error occured on "null/3" level when trying to entering frame
> _WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=False Information:  Was not able to check / back to "calling frame"
> ```


I started a concept which I finally posted here: [$_WD_DEBUG_Calling usage](https://github.com/Danp2/au3WebDriver/pull/362/commits/0d5cf57b48e32e64c0dd57d9194055fa0fc7de65)

Danp2 commented:
> I started a concept which I finally posted here: [$_WD_DEBUG_Calling usage](https://github.com/Danp2/au3WebDriver/pull/362/commits/0d5cf57b48e32e64c0dd57d9194055fa0fc7de65)

I'm having a difficult understanding how this is supposed to work. AFAICS, `$_WD_DEBUG` is set in _WD_FrameList here --

https://github.com/Danp2/au3WebDriver/blob/e2d3160ff38bd95952474da399a1cef2eee8cc9d/wd_helper.au3#L707

and it gets restored here --

https://github.com/Danp2/au3WebDriver/blob/e2d3160ff38bd95952474da399a1cef2eee8cc9d/wd_helper.au3#L765

However, I don't comprehend the need for adding `$_WD_DEBUG_Calling` to __WD_Frame_Internal. Where is `$_WD_DEBUG` being modified in __WD_Frame_Internal such that the following code is needed?

https://github.com/Danp2/au3WebDriver/blob/e2d3160ff38bd95952474da399a1cef2eee8cc9d/wd_helper.au3#L844-L851

mlipok commented:
The reason I used this snippet:
https://github.com/Danp2/au3WebDriver/blob/e2d3160ff38bd95952474da399a1cef2eee8cc9d/wd_helper.au3#L844-L851

Is that this is the only way to get error log from 
https://github.com/Danp2/au3WebDriver/blob/e2d3160ff38bd95952474da399a1cef2eee8cc9d/wd_helper.au3#L853-L855

without this, it will be all the time set `$_WD_DEBUG = $_WD_DEBUG_None` when `__WD_FrameList_Internal` is processing.

mlipok commented:
As I said current version produce such log:
```
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null/3 Information: Error occured on "null/3" level when trying to entering frame
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=False Information:  Was not able to check / back to "calling frame"
```


without this:
https://github.com/Danp2/au3WebDriver/blob/e2d3160ff38bd95952474da399a1cef2eee8cc9d/wd_helper.au3#L844-L851

you will get:

```
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=False Information:  Was not able to check / back to "calling frame"
```

instead

```
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null/3 Information: Error occured on "null/3" level when trying to entering frame
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=False Information:  Was not able to check / back to "calling frame"
```

Danp2 commented:
I think that I understand now. I was originally confuses about the initial value for `$_WD_DEBUG_Calling`. 

FYI, this line is generating an error --

https://github.com/Danp2/au3WebDriver/blob/e2d3160ff38bd95952474da399a1cef2eee8cc9d/wd_demo.au3#L643

mlipok commented:
> FYI, this line is generating an error --
> 
> https://github.com/Danp2/au3WebDriver/blob/e2d3160ff38bd95952474da399a1cef2eee8cc9d/wd_demo.au3#L643


check it again with this mod: [_WD_LoadWait usage](https://github.com/Danp2/au3WebDriver/pull/362/commits/30048eda9085b5b79d5876e2cb612759e3f86caa)

mlipok commented:
after using: [_WD_LoadWait timeout](https://github.com/Danp2/au3WebDriver/pull/362/commits/55d8832ff04c8603cc2a9cdd96ca021677f144eb)

I get:
```
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null/11/0/0 Information: Error occured on "null/11/0/0" level when waiting for a browser page load to complete
__WD_FrameList_Internal ==> Webdriver Exception [10] : Parameters:    Level=null/11
_WD_FrameList ==> Webdriver Exception [10] : Parameters:    ReturnAsArray=True Information:  Was not able to check / back to "calling frame"
```

Danp2 commented:
It still generates an error. Here's the portion of the log where the error occurs --

```
> 640 IsWindowTop = True
__WD_Post: URL=HTTP://127.0.0.1:4444/session/b6e2bebb-6ecb-478b-9007-4cfb74498a6f/frame; Data={"id":null}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":null}
_WD_Window ==> Success [0] : Parameters:   Command=frame   Option={"id":null}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/b6e2bebb-6ecb-478b-9007-4cfb74498a6f/frame; Data={"id":2}
__WD_Post ==> Webdriver Exception [10] : HTTP status = 404 ResponseText={"value":{"error":"no such frame","message":"Unable to locate frame with index: 2","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:188:5\nNoSuchFrameError@chrome://remote/content/shared/webdriver/Errors.jsm:431:5\nswitchToFrame@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:533:15\nreceiveMessage@chrome://remote/content/marionette/actors/MarionetteCommandsChild.jsm:161:31\n"}}
_WD_Window ==> Webdriver Exception [10] : Parameters:   Command=frame   Option={"id":2}
_WD_FrameEnter ==> Webdriver Exception [10] : Parameters:    Identifier=null/2/0 Error on ID#2 > {"id":2}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/b6e2bebb-6ecb-478b-9007-4cfb74498a6f/execute/sync; Data={"script":"return window.top == window.self", "args":[]}
__WD_Post ==> Success [0] : HTTP status = 200 ResponseText={"value":true}
_WD_ExecuteScript ==> Success [0]
_WD_IsWindowTop ==> Success [0]
> 646 IsWindowTop = True
```

mlipok commented:
I need to sleep. I'm going to bed. Will check it ASAP.

mlipok commented:
so the problem is with:
```autoit
; change document context location
_WD_FrameEnter($sSession, 'null/2/0')
```

but it not happen on my side.
please try to add LoadWait
```autoit
; change document context location
_WD_LoadWait($sSession, 1000)
_WD_FrameEnter($sSession, 'null/2/0')
```

Danp2 commented:
IMO adding a call to _WD_LoadWait won't help here because the navigation took place 20 lines earlier and there have already been multiple calls to _WD_FrameList. I suspect that the issue may be that the webpage yields different result because of our different locales. Please compare these results (Ex 1 and Ex 2) to your own --

```
null|||https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe|04dce72b-3e59-4f31-bd48-e60983a638cb
null/0|0|<iframe id="iframeResult" name="iframeResult" allowfullscreen="true" frameborder="0"></iframe>|https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe|5708b7eb-3cea-4200-a96a-6a9db1730e09
null/0/0|0/0|<iframe src="https://www.w3schools.com" title="W3Schools Free Online Web Tutorials"></iframe>|https://www.w3schools.com/|dbd7bf88-1c5a-412c-8af4-c5e5c079c454
null/0/0/0|0/0/0|<iframe src="/howto/tryhow_js_slideshow_ifr.htm" id="howto_iframe"></iframe>|https://www.w3schools.com/howto/tryhow_js_slideshow_ifr.htm|239de0d4-4c94-4020-9315-13d0f141c046
```

![image](https://user-images.githubusercontent.com/7101313/186180542-855f8dcc-d1b6-4fd2-a725-d4ddb8d8859a.png)

mlipok commented:
example 1:
```
null|||https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe|7721ca24-1a5b-46a9-afd0-0154150aa31a
null/0|0|<iframe style="border: 0px none; vertical-align: bottom;" src="https://c58b2aaccd8614bd2c9033016f294e71.safeframe.googlesyndication.com/safeframe/1-0-38/html/container.html" id="google_ads_iframe_/22152718,16833175/sws-hb//w3schools.com//try_it_leaderboard_0" title="3rd party ad content" name="" scrolling="no" marginwidth="0" marginheight="0" data-is-safeframe="true" sandbox="allow-forms allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-top-navigation-by-user-activation" role="region" aria-label="Advertisement" tabindex="0" data-google-container-id="1" data-load-complete="true" width="728" height="90" frameborder="0"></iframe>|about:blank|e453c8bd-1463-463b-a606-99ea8a48803d
null/1|1|<iframe id="iframeResult" name="iframeResult" allowfullscreen="true" frameborder="0"></iframe>|about:blank|09b9c0aa-8176-4e7e-81b0-f8e15b38efbf
null/2|2|<iframe style="display: none;" name="__tcfapiLocator"></iframe>|https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe|ec201f63-2c44-4406-a392-629a9e07ba1b
null/2/0|2/0|<iframe src="https://www.w3schools.com" title="W3Schools Free Online Web Tutorials"></iframe>|https://www.w3schools.com/|c61a6161-1ca3-46f6-82e9-5f8684d5dbb0
null/2/0/0|2/0/0|<iframe src="/howto/tryhow_js_slideshow_ifr.htm" id="howto_iframe"></iframe>|https://www.w3schools.com/howto/tryhow_js_slideshow_ifr.htm|a30cbc7a-c013-420b-8c3e-5b8c83bae650
null/2/0/1|2/0/1|<iframe style="display: none;" name="__tcfapiLocator"></iframe>|about:blank|e133d88d-5cd6-4692-9731-e527562f5c69
null/2/0/2|2/0/2|<iframe style="display: none;" name="__uspapiLocator"></iframe>|about:blank|2c1e3d94-f771-4461-b6fa-958d4622f547
null/3|3|<iframe style="display: none;" name="__uspapiLocator"></iframe>|https://aax-eu.amazon-adsystem.com/s/iu3?cm3ppd=1&d=dtb-pub&csif=t&gdpr=1&gdpr_consent=CPeNQrAPeNQrADlBFAPLCdCsAP_AAH_AACiQI8Nf_X__b3_n-_7___t0eY1f9_7__-0zjhfdt-8N3f_X_L8X_2M7vF36tr4KuR4ku3bBIQdtHOncTUmx6olVrzPsbk2cr7NKJ_Pkmnsbe2dYGH9_n9_z_ZKZ7______7___________________v__________9__________________________________BHgAkw1biALsSxwJtowigRAjCsJDqBQAUUAwtEFhA6uCnZXAT6whYAIBQBGBECHEFGDAIABAIAkIiAkCPBAIgCIBAACABUAhAARsAgsALAwCAAUA0LFGKAIQJCDIgIjlMCAiRIKCeysQSg70NMIQ6ywAoNH_FQgIlACFYGQkLByHBEgJeLJAsxRvkAIwQoBRKgAAA&dl=n-index_n-sharethrough_pm-db5_ym_rbd_n-vmg_ox-db5_smrt_n-nativo_an-db5_3lift&dcc=t|7a826add-8325-4529-9b4a-c0e5af64f89f
null/4|4|<iframe src="https://c58b2aaccd8614bd2c9033016f294e71.safeframe.googlesyndication.com/safeframe/1-0-38/html/container.html" style="visibility: hidden; display: none;"></iframe>|https://ads.pubmatic.com/AdServer/js/user_sync.html?p=159110&gdpr=1&gdpr_consent=CPeNQrAPeNQrADlBFAPLCdCsAP_AAH_AACiQI8Nf_X__b3_n-_7___t0eY1f9_7__-0zjhfdt-8N3f_X_L8X_2M7vF36tr4KuR4ku3bBIQdtHOncTUmx6olVrzPsbk2cr7NKJ_Pkmnsbe2dYGH9_n9_z_ZKZ7______7___________________v__________9__________________________________BHgAkw1biALsSxwJtowigRAjCsJDqBQAUUAwtEFhA6uCnZXAT6whYAIBQBGBECHEFGDAIABAIAkIiAkCPBAIgCIBAACABUAhAARsAgsALAwCAAUA0LFGKAIQJCDIgIjlMCAiRIKCeysQSg70NMIQ6ywAoNH_FQgIlACFYGQkLByHBEgJeLJAsxRvkAIwQoBRKgAAA&us_privacy=1---&predirect=https%3A%2F%2Fu.4dex.io%2Fsetuid%3Fbidder%3Dpubmatic%26gdpr%3D1%26gdpr_consent%3DCPeNQrAPeNQrADlBFAPLCdCsAP_AAH_AACiQI8Nf_X__b3_n-_7___t0eY1f9_7__-0zjhfdt-8N3f_X_L8X_2M7vF36tr4KuR4ku3bBIQdtHOncTUmx6olVrzPsbk2cr7NKJ_Pkmnsbe2dYGH9_n9_z_ZKZ7______7___________________v__________9__________________________________BHgAkw1biALsSxwJtowigRAjCsJDqBQAUUAwtEFhA6uCnZXAT6whYAIBQBGBECHEFGDAIABAIAkIiAkCPBAIgCIBAACABUAhAARsAgsALAwCAAUA0LFGKAIQJCDIgIjlMCAiRIKCeysQSg70NMIQ6ywAoNH_FQgIlACFYGQkLByHBEgJeLJAsxRvkAIwQoBRKgAAA%26uid%3D(PM_UID)|5498a8f6-e8a0-4fc1-bfd1-8b0cf5c3dba5
null/5|5|<iframe style="display: none;" src="https://aax-eu.amazon-adsystem.com/s/iu3?cm3ppd=1&amp;d=dtb-pub&amp;csif=t&amp;gdpr=1&amp;gdpr_consent=CPeNQrAPeNQrADlBFAPLCdCsAP_AAH_AACiQI8Nf_X__b3_n-_7___t0eY1f9_7__-0zjhfdt-8N3f_X_L8X_2M7vF36tr4KuR4ku3bBIQdtHOncTUmx6olVrzPsbk2cr7NKJ_Pkmnsbe2dYGH9_n9_z_ZKZ7______7___________________v__________9__________________________________BHgAkw1biALsSxwJtowigRAjCsJDqBQAUUAwtEFhA6uCnZXAT6whYAIBQBGBECHEFGDAIABAIAkIiAkCPBAIgCIBAACABUAhAARsAgsALAwCAAUA0LFGKAIQJCDIgIjlMCAiRIKCeysQSg70NMIQ6ywAoNH_FQgIlACFYGQkLByHBEgJeLJAsxRvkAIwQoBRKgAAA&amp;dl=n-index_n-sharethrough_pm-db5_ym_rbd_n-vmg_ox-db5_smrt_n-nativo_an-db5_3lift"></iframe>|https://c58b2aaccd8614bd2c9033016f294e71.safeframe.googlesyndication.com/safeframe/1-0-38/html/container.html|3a16f4be-e04d-4fe9-a010-59e7c7caa3ac
null/6|6|<iframe id="adg-0-sync" marginwidth="0" marginheight="0" scrolling="no" src="https://ads.pubmatic.com/AdServer/js/user_sync.html?p=159110&amp;gdpr=1&amp;gdpr_consent=CPeNQrAPeNQrADlBFAPLCdCsAP_AAH_AACiQI8Nf_X__b3_n-_7___t0eY1f9_7__-0zjhfdt-8N3f_X_L8X_2M7vF36tr4KuR4ku3bBIQdtHOncTUmx6olVrzPsbk2cr7NKJ_Pkmnsbe2dYGH9_n9_z_ZKZ7______7___________________v__________9__________________________________BHgAkw1biALsSxwJtowigRAjCsJDqBQAUUAwtEFhA6uCnZXAT6whYAIBQBGBECHEFGDAIABAIAkIiAkCPBAIgCIBAACABUAhAARsAgsALAwCAAUA0LFGKAIQJCDIgIjlMCAiRIKCeysQSg70NMIQ6ywAoNH_FQgIlACFYGQkLByHBEgJeLJAsxRvkAIwQoBRKgAAA&amp;us_privacy=1---&amp;predirect=https%3A%2F%2Fu.4dex.io%2Fsetuid%3Fbidder%3Dpubmatic%26gdpr%3D1%26gdpr_consent%3DCPeNQrAPeNQrADlBFAPLCdCsAP_AAH_AACiQI8Nf_X__b3_n-_7___t0eY1f9_7__-0zjhfdt-8N3f_X_L8X_2M7vF36tr4KuR4ku3bBIQdtHOncTUmx6olVrzPsbk2cr7NKJ_Pkmnsbe2dYGH9_n9_z_ZKZ7______7___________________v__________9__________________________________BHgAkw1biALsSxwJtowigRAjCsJDqBQAUUAwtEFhA6uCnZXAT6whYAIBQBGBECHEFGDAIABAIAkIiAkCPBAIgCIBAACABUAhAARsAgsALAwCAAUA0LFGKAIQJCDIgIjlMCAiRIKCeysQSg70NMIQ6ywAoNH_FQgIlACFYGQkLByHBEgJeLJAsxRvkAIwQoBRKgAAA%26uid%3D(PM_UID)" style="border: 0px; display: none;" width="0" height="0" frameborder="0"></iframe>|https://c58b2aaccd8614bd2c9033016f294e71.safeframe.googlesyndication.com/safeframe/1-0-38/html/container.html|52378401-95ca-4dad-b52a-97d2f42ac5f8
null/6/0|6/0|<iframe title="Blank" src="https://googleads.g.doubleclick.net/xbbe/pixel?d=CJyntKMDEJnIm6UDGKSRo9EBMAE&amp;v=APEucNVSlScLKAb7SYml5EFar7xcJXhsNn-H9yl0sCwHTYVKm-sYWAn3rORBbGwvypNYNMSGjH8-oCGSY63nMLj0_v8Oi5xV63GNBgtT07G4fp87vsi1bCCuOAWObkRRqq3vqWgAFcLXlU7jBOGs6osBGKo8uSW2S7eXVq97etfBWeUVCLkz7MNwOYmjG2GhhLAUeNS08P3k1Kw7kz4WhlXpazzDnBAOuA1eWBhqr00BkHr9lMl5oByMPu19uRn4uhShnq9hVXsQvDkeFk5HXZYPBAl9ygIi_TgLfLT-ZqxdgPgLGeHiP6owPhlQXTz1BuA01CChdjpvuxhdkWsTfs90bpW13rxuY0l8gQxlcNwewz81PCtzw1iNhbkbu2IuJKsl5gdCVRZDi-FWfYGO1jC1gFIvZqdxy5rga9hmLiBjpDZQG8hvpPYoobCNmfz5yEzLY24ja-KZUOrUnniroL3f2i3rMuJxRRni2c6EPTGCUGfRWDuCOx2D3hlVZl-78LkcMVuPbQ8klqlGt4sqKgTc2HXuteBT1_02X63urjF3Lx5FbwaP5NRJQYLUgWyh-LLgUqqB56BYtWo3DSyeR3ggjzgMSH3TCVuOn9zm1MAp3a_j-esZWooDm1Asu_4evHb2zeImmY5f8T5t8cPJrjhGPyrbGQaeWR29OFmcmMa3UFqXavn9bOHK9sNm4D72TCQqs-Nq9BmIxjG3gHedYCmjqvWsFUfSfnJ1vw0SF8rwSRvaxhNeE_1WAP4ceZLRKO3lPpAhjbxJ_bftzMAOfsX2xjBSWVGaGZtSha2yjqXGIDtjJdeqNrq1KJhD_FkcnrnBelLnyaKMI1JxYGFF6jVi7dJ8vspS0MJkzAEd_W2voazRfTxCYIsbChTDHANME7MNzm9D8S-YMRRIKW0DBGfg0KgNMxPI3nFwbJk99RiNB0e_ipaZ5IecUwRPeUKJWc1EOUKkr_sD3oKGsEIHBv6_2DfJ-P5IFhdafmTFuEwzmxQ5u24O7qmpd7FiiR0AFoX7yCfsGP5nhYZd6DT4ghoND29YXQo1ihFOQ5pE5A7oGMQb7HEgm9TtETQpMkaSg8VGd4qa5ke5uDBJp8I3H4xWy1zn38GyPzTYfeVeXtgA6oTkq1BBtT3tOpkE943xkdRPOeYVxAa8E3rO_IjyQxb7XIk_hSoRB8h9fTDj6Ci_vj32vRBusQFzFArybk-ukwv5cuTyFGvOUFYVrSrfPfW3-vGDxRZLH-TeoCjsM_OreBRj7v-4TZ4cA8SiRQzWkCPa1HFdTn8_CKHotSQwGyfqyx2zcTP7y7o2hbjw7MOLgDkZXRpEF-dXJaFtuLxjf3kbpDKRncxNyTPPN3RtAtv_h4XSxtKQLXHg8orgsNDv9lYfcESCgxuQk3u2dN6q1bEVhE_oj2YfgCNs5WjUe8hoxKyjsItzeexC97hwlad_rNCp1tO3x854vVXjHZkdN3AYPJsxl6RRcilv6XvjC6xOE4u1qGGapqQZZqkSUbXkt2NNTTpKqO97ynxqHMTwp9ccXNTNyjCYShBsRp89XSE_wtT1PDtx9SDCW0QuYgq4Wl5bRQ1nSgENH2K0_4Nn8SsKR4WhRhHnyHmxiu5BbM5DsMG6dCQ66EzMQUTWGrOSafhGRvcm1p8Zau_qNVUbylvVkG_4UE8BvSJrA8yTmcfvcdUnUwH-mzySX7VKbDKgKzTA1Q1_IFvz0bLWoUL4mSFMuTKP36h_BeEmIm3Ia2mVkt-gaoUQ0jxLtMCqpG0DFY8I0GmctGR2qOS1OiNL4USZ8PVeMDVoUsPtnUxjloXmT4ThX3oyuIpeR12ZyXeEFYLPVwQHw1rto8Vmjh0XrYfwHuDahIpfiLgs8fGYfM8CxAI2Q6Dj3A4MtFGXMeTBH1Kdpr5DqBpR1LGC1sSBECRGSacC4CZ8U8h0R7F2pcQ-KYaixLmf-QUGLxZYu1g7kioT3ppjw_Bbc-fSkrlT2OkgQTJQflcByvltDb5RhF5OIm1vE8_VUCa7kHMsKo3sRtLJi6QQTPLFNcXkcNfgIvTbWuYynYvSIIF-q_bF9n6jvLmB3k9bHdPgDpTT4UyKOXk0DhnejdLBmmz0UjXpJIPWn1y4sj5XciDYq63jId0XUEDsZmUjpmB6TwBHoAHl4wS4sOxgV-1jHtq2oOjwiG52R6eJOTgu6duIEQtURjLzyCUWq35Ugy85-S1k62WMs5feD8REdVin9ocl9QNLUWm2hK-ymv-xQeGhlSjuoZPD1iZOL09_oAtNi32j31Vc6bfUCOOpIabBD1RnTPWRJ9z8yjXyaNsZ0hQJBFm0JecIvAxM2dHgTZIjcjSxyvFyZVYITxntHeUPVm5sTRNmi95K6NjVv1MXLVrNTV8OE6Vp0IBBiibQF2BfNm88K9iVZfcQLvPR72-iH6UT0sfKqMoZBGYbMbUO5R1DjYbI-6AaTxAbzyuLJX4V6iVRpe7j6gfA5KXuA4y3LSGLjeClXSVH1CqsB7zkAd00t-_TRY2XpsqSSd9YplPO8bDjH2K9UPlPD_JJWz5O0379NxBBnbBkH194o4nvYf1UhtqBIbRocnqs09oc9WJQHLIDS4A8Wh9fsOQk3vkKElzhfP2fUgawiWa2WGI47Wh-Um0iRmIcG4W5hHRf925U8FbptTXgiFk1zQYHuOJjO4X7KGlpv7j0yEY1-Loe6YY7cdEtwB7RdvntI3Zv4r35lSdWkAwTYrCM6X30Xb-xujgGQe5syu6SJPalyi6CZw0CwzdIYdUqII-bGETcdVHA5EPnE6m_6vf1o7MFiE1ysrIo7xDCT0zEUB6YkNpsRWZet6clYN7KBY0hTK0aXmOzJRPZpyDV-POOTGEpdjRXteSA9RI83fzW8qdCiO1HGgfCzB_9y7bgZjIHxV4MiDJflUEG05j_h2z98MZQBPt_r96AcTYajx2dF0jw017IluedVlupQdHEUgfH8sX11TOAMJbve-6nRRb9wSP-G5hrEQQMxi_rRllmvE3BTDTkBKDbz-vtlbjm5JOSR537F9WGGO6UZ7frKXgNAJcSjASbxn22ZDTlbKFOiwryEBa19Ep3yLQrzMTV-f39bEe09OTDsloJXpa0O4HgMMHtpd2AeAKbpEjcdeH8hsMofgoom5WXSW41wqt_se1NYdCP1MquKwBHFS0zMvESyzMA0d2bnSJnndlj0T-FQeHScrtPWsOyUm1DzffEo4gBrpI1duygKkTM9iq1NXPyODwVzohg1UIdYcWoyIsIXgKt78XvH0aF3pI1sTatDCLsnomZQxw5GE4t3q6RBCXA3uifb9E" style="display:none" aria-hidden="true"></iframe>|https://googleads.g.doubleclick.net/xbbe/pixel?d=CJyntKMDEJnIm6UDGKSRo9EBMAE&v=APEucNVSlScLKAb7SYml5EFar7xcJXhsNn-H9yl0sCwHTYVKm-sYWAn3rORBbGwvypNYNMSGjH8-oCGSY63nMLj0_v8Oi5xV63GNBgtT07G4fp87vsi1bCCuOAWObkRRqq3vqWgAFcLXlU7jBOGs6osBGKo8uSW2S7eXVq97etfBWeUVCLkz7MNwOYmjG2GhhLAUeNS08P3k1Kw7kz4WhlXpazzDnBAOuA1eWBhqr00BkHr9lMl5oByMPu19uRn4uhShnq9hVXsQvDkeFk5HXZYPBAl9ygIi_TgLfLT-ZqxdgPgLGeHiP6owPhlQXTz1BuA01CChdjpvuxhdkWsTfs90bpW13rxuY0l8gQxlcNwewz81PCtzw1iNhbkbu2IuJKsl5gdCVRZDi-FWfYGO1jC1gFIvZqdxy5rga9hmLiBjpDZQG8hvpPYoobCNmfz5yEzLY24ja-KZUOrUnniroL3f2i3rMuJxRRni2c6EPTGCUGfRWDuCOx2D3hlVZl-78LkcMVuPbQ8klqlGt4sqKgTc2HXuteBT1_02X63urjF3Lx5FbwaP5NRJQYLUgWyh-LLgUqqB56BYtWo3DSyeR3ggjzgMSH3TCVuOn9zm1MAp3a_j-esZWooDm1Asu_4evHb2zeImmY5f8T5t8cPJrjhGPyrbGQaeWR29OFmcmMa3UFqXavn9bOHK9sNm4D72TCQqs-Nq9BmIxjG3gHedYCmjqvWsFUfSfnJ1vw0SF8rwSRvaxhNeE_1WAP4ceZLRKO3lPpAhjbxJ_bftzMAOfsX2xjBSWVGaGZtSha2yjqXGIDtjJdeqNrq1KJhD_FkcnrnBelLnyaKMI1JxYGFF6jVi7dJ8vspS0MJkzAEd_W2voazRfTxCYIsbChTDHANME7MNzm9D8S-YMRRIKW0DBGfg0KgNMxPI3nFwbJk99RiNB0e_ipaZ5IecUwRPeUKJWc1EOUKkr_sD3oKGsEIHBv6_2DfJ-P5IFhdafmTFuEwzmxQ5u24O7qmpd7FiiR0AFoX7yCfsGP5nhYZd6DT4ghoND29YXQo1ihFOQ5pE5A7oGMQb7HEgm9TtETQpMkaSg8VGd4qa5ke5uDBJp8I3H4xWy1zn38GyPzTYfeVeXtgA6oTkq1BBtT3tOpkE943xkdRPOeYVxAa8E3rO_IjyQxb7XIk_hSoRB8h9fTDj6Ci_vj32vRBusQFzFArybk-ukwv5cuTyFGvOUFYVrSrfPfW3-vGDxRZLH-TeoCjsM_OreBRj7v-4TZ4cA8SiRQzWkCPa1HFdTn8_CKHotSQwGyfqyx2zcTP7y7o2hbjw7MOLgDkZXRpEF-dXJaFtuLxjf3kbpDKRncxNyTPPN3RtAtv_h4XSxtKQLXHg8orgsNDv9lYfcESCgxuQk3u2dN6q1bEVhE_oj2YfgCNs5WjUe8hoxKyjsItzeexC97hwlad_rNCp1tO3x854vVXjHZkdN3AYPJsxl6RRcilv6XvjC6xOE4u1qGGapqQZZqkSUbXkt2NNTTpKqO97ynxqHMTwp9ccXNTNyjCYShBsRp89XSE_wtT1PDtx9SDCW0QuYgq4Wl5bRQ1nSgENH2K0_4Nn8SsKR4WhRhHnyHmxiu5BbM5DsMG6dCQ66EzMQUTWGrOSafhGRvcm1p8Zau_qNVUbylvVkG_4UE8BvSJrA8yTmcfvcdUnUwH-mzySX7VKbDKgKzTA1Q1_IFvz0bLWoUL4mSFMuTKP36h_BeEmIm3Ia2mVkt-gaoUQ0jxLtMCqpG0DFY8I0GmctGR2qOS1OiNL4USZ8PVeMDVoUsPtnUxjloXmT4ThX3oyuIpeR12ZyXeEFYLPVwQHw1rto8Vmjh0XrYfwHuDahIpfiLgs8fGYfM8CxAI2Q6Dj3A4MtFGXMeTBH1Kdpr5DqBpR1LGC1sSBECRGSacC4CZ8U8h0R7F2pcQ-KYaixLmf-QUGLxZYu1g7kioT3ppjw_Bbc-fSkrlT2OkgQTJQflcByvltDb5RhF5OIm1vE8_VUCa7kHMsKo3sRtLJi6QQTPLFNcXkcNfgIvTbWuYynYvSIIF-q_bF9n6jvLmB3k9bHdPgDpTT4UyKOXk0DhnejdLBmmz0UjXpJIPWn1y4sj5XciDYq63jId0XUEDsZmUjpmB6TwBHoAHl4wS4sOxgV-1jHtq2oOjwiG52R6eJOTgu6duIEQtURjLzyCUWq35Ugy85-S1k62WMs5feD8REdVin9ocl9QNLUWm2hK-ymv-xQeGhlSjuoZPD1iZOL09_oAtNi32j31Vc6bfUCOOpIabBD1RnTPWRJ9z8yjXyaNsZ0hQJBFm0JecIvAxM2dHgTZIjcjSxyvFyZVYITxntHeUPVm5sTRNmi95K6NjVv1MXLVrNTV8OE6Vp0IBBiibQF2BfNm88K9iVZfcQLvPR72-iH6UT0sfKqMoZBGYbMbUO5R1DjYbI-6AaTxAbzyuLJX4V6iVRpe7j6gfA5KXuA4y3LSGLjeClXSVH1CqsB7zkAd00t-_TRY2XpsqSSd9YplPO8bDjH2K9UPlPD_JJWz5O0379NxBBnbBkH194o4nvYf1UhtqBIbRocnqs09oc9WJQHLIDS4A8Wh9fsOQk3vkKElzhfP2fUgawiWa2WGI47Wh-Um0iRmIcG4W5hHRf925U8FbptTXgiFk1zQYHuOJjO4X7KGlpv7j0yEY1-Loe6YY7cdEtwB7RdvntI3Zv4r35lSdWkAwTYrCM6X30Xb-xujgGQe5syu6SJPalyi6CZw0CwzdIYdUqII-bGETcdVHA5EPnE6m_6vf1o7MFiE1ysrIo7xDCT0zEUB6YkNpsRWZet6clYN7KBY0hTK0aXmOzJRPZpyDV-POOTGEpdjRXteSA9RI83fzW8qdCiO1HGgfCzB_9y7bgZjIHxV4MiDJflUEG05j_h2z98MZQBPt_r96AcTYajx2dF0jw017IluedVlupQdHEUgfH8sX11TOAMJbve-6nRRb9wSP-G5hrEQQMxi_rRllmvE3BTDTkBKDbz-vtlbjm5JOSR537F9WGGO6UZ7frKXgNAJcSjASbxn22ZDTlbKFOiwryEBa19Ep3yLQrzMTV-f39bEe09OTDsloJXpa0O4HgMMHtpd2AeAKbpEjcdeH8hsMofgoom5WXSW41wqt_se1NYdCP1MquKwBHFS0zMvESyzMA0d2bnSJnndlj0T-FQeHScrtPWsOyUm1DzffEo4gBrpI1duygKkTM9iq1NXPyODwVzohg1UIdYcWoyIsIXgKt78XvH0aF3pI1sTatDCLsnomZQxw5GE4t3q6RBCXA3uifb9E|2af06903-db67-40e1-8b85-68a1d2684a7e
null/6/1|6/1|<iframe src="//tpc.googlesyndication.com/sodar/Enqz_20U.html" style="display: none;" width="0" height="0"></iframe>|https://tpc.googlesyndication.com/sodar/Enqz_20U.html|703b0652-6152-49ef-83a2-ad6be91c5155
null/7|7|<iframe id="adg-1-sync" marginwidth="0" marginheight="0" scrolling="no" src="https://secure-assets.rubiconproject.com/utils/xapi/multi-sync.html?p=onfocus&amp;endpoint=eu" style="border: 0px; display: none;" width="0" height="0" frameborder="0"></iframe>|https://eus.rubiconproject.com/usync.html?p=onfocus&endpoint=eu|45de0b61-67b7-48e4-bbef-4e356a951f1f
null/8|8|<iframe src="https://www.google.com/recaptcha/api2/aframe" style="display: none;" width="0" height="0"></iframe>|https://www.google.com/recaptcha/api2/aframe|9768b0af-150e-4c33-b49e-e7209168c60c
null/9|9|<iframe src="https://google-bidout-d.openx.net/w/1.0/pd?plm=5" style="display:none;" width="0" height="0"></iframe>|https://google-bidout-d.openx.net/w/1.0/pd?plm=5|1ed9fce5-1cc6-4e8a-8cdb-7118055464cc
```

![image](https://user-images.githubusercontent.com/11089482/186383247-b4777690-a47e-49cb-9810-b5cda2f325c8.png)

mlipok commented:
so we should change 
```autoit
_WD_FrameEnter($sSession, 'null/2/0')
```
to:
```autoit
_WD_FrameEnter($sSession, 'null/0')
```

mlipok commented:
This site is very strange because

I had to stop checking for error after LoadWait
```autoit
_WD_LoadWait($sSession, 100, 10 * 1000)
; intentionally do not check for error here

$iErr = @error
If 0 And @error Then
	$sMessage = 'Error occured on "' & $sLevel & '" level when waiting for a browser page load to complete'
Else
```

because Ex1 was always succesfull but Ex2 in most cases end due to LoadWait timeout.

After I stop checking I get such results:

![image](https://user-images.githubusercontent.com/11089482/186392069-c7e78a50-42ad-4211-9278-0ebb1f103b21.png)

mlipok commented:
I think we should provide new feature for `_WD_LoadWait()` as there can be cases that you would like to wait not for `complete` but for `interactive` or even `loaded` will be enough to continue processing like there was no error.

Reference:
https://www.w3schools.com/jsref/prop_doc_readystate.asp
```
uninitialized - Has not started loading
loading - Is loading
loaded - Has been loaded
interactive - Has loaded enough to interact with
complete - Fully loaded
```

Related:
https://www.autoitscript.com/forum/topic/208640-webdriver-udf-help-support-iv/?do=findComment&comment=1506254

If you agree then I open separate Issue for this.

mlipok commented:
> it would be helpfull to change _WD_LoadWait behavior on the fly without touching capabilities.
> If you agree then I open separate Issue for this.

as so far this https://github.com/Danp2/au3WebDriver/pull/384 is only error handling improvements so I still wait on your decision about `on the fly`  feature.

Danp2 commented:
@mlipok

- I created a separate issue (see #385) to discuss your proposed enhancements to _WD_LoadWait. 
- Why do you need to call _WD_LoadWait inside __WD_FrameList_Internal?

mlipok commented:
> * I created a separate issue (see [_WD_LoadWait() enhancements #385](https://github.com/Danp2/au3WebDriver/issues/385)) to discuss your proposed enhancements to _WD_LoadWait.

Thanks

> * Why do you need to call _WD_LoadWait inside __WD_FrameList_Internal?

Because `_WD_LoadWait` which is used after navigation is not checking for all frames but only on top document.

mlipok commented:
For this reason currently `_WD_LoadWait` is required and should be refactored/enchanged in `_WD_FrameList`.

But for this stage please leave it as it is.
And if you do not have other questions or request, please to not hesitate to merge this PR.

Danp2 commented:
>>   Why do you need to call _WD_LoadWait inside __WD_FrameList_Internal?

> Because _WD_LoadWait which is used after navigation is not checking for all frames but only on top document.

But is it still required? Please think before answering. Previous revisions of _WD_FrameList returned more details about the frame, such as HTML. Was the need for calling _WD_LoadWait for each frame eliminated when those features were removed?

mlipok commented:
results when I use `_WD_LoadWait`

![image](https://user-images.githubusercontent.com/11089482/186440997-b1a60792-3ae8-4750-81db-7b665e906f18.png)

![image](https://user-images.githubusercontent.com/11089482/186441117-e52d2b34-8dc3-449b-b36b-4c0f23eaea93.png)


results when I comment out `_WD_LoadWait`

![image](https://user-images.githubusercontent.com/11089482/186441284-115cf7fc-8e34-48f0-8565-be3e4be689f7.png)

![image](https://user-images.githubusercontent.com/11089482/186441379-865d9af1-4214-4f53-813d-b9948db46304.png)

Danp2 commented:
> I had to stop checking for error after LoadWait because Ex1 was always succesfull but Ex2 in most cases end due to LoadWait timeout.

You should dig into this further to better understand the reason for the difference in behavior.

> If 0 And @error Then

This is ugly "fix"

mlipok commented:
I have added `MsgBox`
```autoit
MsgBox(0, "", 'Slow down internet speed to "Regular 3G" or even "Regular 2G"')
; now lets try to check frame list and using locations as path:
; go to website
_WD_Navigate($sSession, 'https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_iframe')
_WD_LoadWait($sSession)
```
and changed in `network tab` in `DevTools`:
![image](https://user-images.githubusercontent.com/11089482/186443722-da84c333-5a09-4d14-95da-70086b286e9b.png)

And I get
![image](https://user-images.githubusercontent.com/11089482/186443954-4c1e5488-f698-4c0f-816a-5e455f49b023.png)

then I wait 10 second, and close _ArrayDisplay result from `Example 2`

And I get _ArrayDisplay result form `Example 3`
![image](https://user-images.githubusercontent.com/11089482/186444020-bacd9a2e-48ee-4c0d-aad6-6955b25cbcd1.png)


They are different as when I wait 10 seconds then document was still loading.
For this reason `_WD_LoadWait` is required IMHO.

mlipok commented:
> @mlipok Those screenshots aren't even from the same examples.

my mistake. sorry. take a look for my recent 2 screenshots.

Danp2 commented:
>  For this reason _WD_LoadWait is required IMHO.

Ok... that makes sense when you think about it, because you are wanting a complete list of all frames so you have to allow the frames to load in case they contain other frames.

> You should dig into this further to better understand the reason for the difference in behavior.
> This is ugly "fix"

Wouldn't it make more sense here to detect a timeout and ignore that?

mlipok commented:
> Wouldn't it make more sense here to detect a timeout and ignore that?

Good point

mlipok commented:
> Wouldn't it make more sense here to detect a timeout and ignore that?

Please take a look on:
[LoadWait is error <> $_WD_ERROR_Timeout](https://github.com/Danp2/au3WebDriver/pull/362/commits/6e41f1512079ac3b037aaf7f5dc89fe9fc18e582)
and
[_WD_LoadWait $iErr](https://github.com/Danp2/au3WebDriver/pull/362/commits/2eaab01564f9016ad37fbad795c063d04249457a)

mlipok commented:
As TODO in the future:

_WD_FrameList should return informationa about visibility of the frame

for example:
https://stackoverflow.com/questions/19669786/check-if-element-is-visible-in-dom

![image](https://user-images.githubusercontent.com/11089482/186519722-2d7d707a-e1d4-49a4-af2b-92a05e6c4595.png)
